kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/GwtComponentServiceImpl.java

```diff
@@ -244,12 +250,20 @@ public GwtConfigComponent findWireComponentConfigurationFromPid(GwtXSRFToken xsr
                 if (conf != null) {
                     conf.getConfigurationProperties().put(ConfigurationAdmin.SERVICE_FACTORYPID, factoryPid);
                 }
-                if ((conf != null) && (conf.getDefinition() == null)) {
+                if (conf != null && conf.getDefinition() == null) {
                     String temporaryName = String.valueOf(System.nanoTime());
                     cs.createFactoryConfiguration(factoryPid, temporaryName, extraProps, false);
                     try {
-                        // wait for the services to be up
-                        TimeUnit.MILLISECONDS.sleep(500);
+
+                        // track and wait for the wire Component
+                        final BundleContext bundleContext = FrameworkUtil.getBundle(this.getClass()).getBundleContext();
+                        String filterString = "(" + ConfigurationService.KURA_SERVICE_PID + "=" + temporaryName + ")";
+                        Filter filter = bundleContext.createFilter(filterString);
+                        final ServiceTracker wireComponentTracker = new ServiceTracker(bundleContext, filter, null);
+                        wireComponentTracker.open();
+                        wireComponentTracker.waitForService(SERVICE_WAIT_TIMEOUT);
+                        wireComponentTracker.close();
+
                         conf = cs.getComponentConfiguration(temporaryName);
```

Comment:
You can also do one more thing for enhancement.
```
T serviceInstance = (T) serviceTracker.waitForService(WAIT_FOR_SERVICE_SEC * 1000);
if (serviceInstance != null) { .. then invoke conf = cs.getComponentConfiguration(temporaryName); ...
```

I believe it's more compact this way.